### DISCUSSION TITLE
Middleware / API routes etc

### DISCUSSION THREAD

**Discussion by deckchairlabs** [exhibitionist-digital/ultra#98] [Ideas]
So we probably have two options for middleware.

1. Recommend using the Oak server integration `src/oak/handler.ts`

2. Write, borrow our own routing/middleware solution on top of `std/serve`

**Comment by deckchairlabs** [comment]
I'm pretty behind the first option still. Option 2 just feels like we'd be writing code, because we can? I think we need to list pros and cons for both approaches.

Ultra for me has always been about streaming and rendering, nothing more nothing less.

**Comment by dburles** [comment]
There's definitely pros/cons to both. There was some reason originally that @mashaal dropped Oak (I can't recall why that was). Writing things ourselves has advantages if the server requirements for Ultra are fairly minimal and we don't require everything that Oak offers, we can eliminate a large dependency and have tighter control.

**Comment by mashaal** [comment]
I think ideally we have something like:

- Oak handler (done), which means people can easily use Ultra in an Oak project if they want
- Deprecate API routes in their current form, as they aren't fully inline with how the rest of Ultra works, and they are nqr
- Implement some 'simple' middleware which lets custom handling of requests. This could be used for custom non-ultra responses (API routes) or per request updates to the Ultra rendering options (setting `lang`), etc

Again, that is very 'Oak' like, but the reasons we moved around from Oak were that static files in Oak a bit of a black box, and at the time didn't work well with Deno Deploy (not sure they still do) https://github.com/oakserver/oak/issues/416#issuecomment-997476494 ...and we ended up creating our own static file handler anyway. It just seemed like a bit overkill for a simple 'barebones' project.

Deploying Ultra to various providers is still a priority, so we need to imagine how these changes will effect that.

**Reply by Industrial** [reply]
I got some inspiration today and tried implementing a middleware handler https://github.com/exhibitionist-digital/ultra/pull/100

@dburles, @mashaal, @deckchairlabs: What do you think about this implementation?

**Reply by deckchairlabs** [reply]
Yeah, I think if we just change the implementation to follow the norm of `server.use(middleware)` this should be more than ideal. I'm thinking we could then refactor a lot of the routing logic in requestHandler to be internal middleware.

**Reply by Industrial** [reply]
I did just that. https://github.com/exhibitionist-digital/ultra/pull/109
